refactor: clean up _wcpay_feature_tokenized_cart_ece feature

frosso opened the pull request:
Fixes #9725

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

Now that the ECE with Tokenized Cart is released and stable, we can clean up the codebase by removing obsolete feature flags and any remaining legacy implementations.

Here's what I did on this PR:
- Moved files from `client/tokenized-express-checkout` to `client/express-checkout` (you'll see some changes aren't explicitly marked as "moved" - like in `client/express-checkout/event-handlers.js` )
- Removed the (now unused) utility methods in `client/checkout/api/index.js` and the API handlers defined in `includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php`
- Removed the feature flag in `WC_Payments_Features::is_tokenized_cart_ece_enabled()` / `_wcpay_feature_tokenized_cart_ece` and everywhere they were checked
- Moved (or removed) some AJAX-related methods from `includes/express-checkout/class-wc-payments-express-checkout-button-helper.php` into `includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php`
- Added tests in `tests/unit/express-checkout/test-class-wc-payments-express-checkout-ajax-handler.php` for HK-based addresses, states, etc.
- Deleted the `wcpayPaymentRequestParams` global in favor of `wcpayExpressCheckoutParams`

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

There are a few areas to check: checkout capability and settings.

1️⃣ Settings (e-z)
- Navigate to WooCommerce > Settings > Payments > WooPayments
- Click on "Customize", next to "Google Pay / Apple Pay"
- The "Preview" should still work (no change compared to `develop`)
<img width="894" alt="Screenshot 2025-05-20 at 5 16 56 PM" src="https://github.com/user-attachments/assets/bc6796c3-328c-455d-a4a2-894f96ee25a0" />

2️⃣ Checkout
- Ensure you have Google Pay / Apple Pay enabled in the settings
- Navigate to the product page of a physical product
- Add the product to the cart
- Go to the cart (or checkout) page
- Click on the Google Pay button
- Interact with the shipping methods: the totals should change according to the shipping region you selected
- Click on "Pay now" on the payment sheet
<img width="589" alt="Screenshot 2025-05-20 at 5 20 25 PM" src="https://github.com/user-attachments/assets/775554ab-4501-4337-b2d9-e0bef8bb4b18" />

- You should be landed on the "Order received" page

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

frosso committed:
chore: clean up _wcpay_feature_tokenized_cart_ece feature

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10784` or branch name `chore/remove-deprecated-non-tokenized-ece` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=chore/remove-deprecated-non-tokenized-ece) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 88f7abbce85229e3978c60773a12c0f75f20e677
- Build time: 2025-05-29 16:17:35 UTC

_Note: the build is updated when a new commit is pushed to this PR._

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
WIP

frosso commented on client/express-checkout/blocks/components/express-checkout-preview.js:
```diff
@@ -81,7 +84,7 @@ export const ExpressCheckoutPreviewComponent = ( {
 			<div
 				key={ `${ buttonType }-${ theme }` }
-				style={ { minHeight: `${ height }px` } }
+				style={ { minHeight: `${ height }px`, width: '100%' } }
```
Decided against this, I think they're inline to prevent overrides.

frosso committed:
Merge branch 'develop' into chore/remove-deprecated-non-tokenized-ece

frosso committed:
arrow function w/ consistent body

frosso committed:
more cleanup

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
tests refactor & relocation

frosso commented on client/express-checkout/blocks/index.js:
```diff
@@ -11,7 +11,7 @@ import { getConfig } from 'wcpay/utils/checkout';
 import ExpressCheckoutContainer from './components/express-checkout-container';
 import { checkPaymentMethodIsAvailable } from '../utils/checkPaymentMethodIsAvailable';
 
-const expressCheckoutElementApplePay = ( api ) => ( {
```
All the changes here are coming from `client/tokenized-express-checkout/blocks/index.js`

frosso commented on client/express-checkout/event-handlers.js:
```diff
@@ -3,69 +3,94 @@
  * External dependencies
  */
 import { __ } from '@wordpress/i18n';
+import { applyFilters } from '@wordpress/hooks';
```
All the changes here are coming from `client/tokenized-express-checkout/event-handlers.js`

frosso commented on client/express-checkout/index.js:
```diff
@@ -1,17 +1,23 @@
+ */
 import { __ } from '@wordpress/i18n';
-import { debounce } from 'lodash';
+import { addAction, removeAction } from '@wordpress/hooks';
```
All the changes here are coming from `client/tokenized-express-checkout/index.js`

frosso committed:
Merge branch 'develop' into chore/remove-deprecated-non-tokenized-ece

frosso committed:
Merge remote-tracking branch 'origin/develop' into chore/remove-deprecated-non-tokenized-ece

mgascam reviewed (commented):
Thanks for working on this cleanup, @frosso! I followed the testing instructions and did a bit of exploratory testing — everything worked as expected and I didn't encounter any issues.

I also reviewed the code changes. Everything looked good on a first pass, although I focused on the files where the changes were smaller or easier to follow. I plan to do a second pass focusing on the files with more substantial changes.

htdat commented:
@frosso - should we also clean up the WCPay Dev tool for the similar `Enable Cart-Token implementation for ECE` flag?

frosso commented:
@htdat yep, let's do it! I just pinged you on the dev tools PR :)

frosso committed:
Merge branch 'develop' into chore/remove-deprecated-non-tokenized-ece

frosso committed:
Merge branch 'develop' into chore/remove-deprecated-non-tokenized-ece

htdat commented on includes/express-checkout/class-wc-payments-express-checkout-button-display-handler.php:
```diff
@@ -84,9 +84,11 @@ public function init() {
 		$is_payment_request_enabled = 'yes' === $this->gateway->get_option( 'payment_request' );
 
-		if ( $is_woopay_enabled || $is_payment_request_enabled ) {
+		if ( $is_woopay_enabled ) {
 			add_action( 'wc_ajax_wcpay_add_to_cart', [ $this->express_checkout_ajax_handler, 'ajax_add_to_cart' ] );
+		}
 
+		if ( $is_woopay_enabled || $is_payment_request_enabled ) {
```
What's the rationale for this change? This might break Google Pay / Apple Pay?

frosso replied:
This `init` method was initializing the "add to cart" AJAX handler, which was used by both the non-tokenized ECE and WooPay.
But by cleaning up the non-tokenized ECE and only having the tokenized ECE in the codebase (which leverages the Store API to add a product to the cart), we don't need the `wc_ajax_wcpay_add_to_cart` handler anymore.
The `wc_ajax_wcpay_add_to_cart` is only used for WooPay.

Then, the rest of the actions in this method (like `woocommerce_after_add_to_cart_form` / `woocommerce_proceed_to_checkout` / etc.)  are still needed for both WooPay and the Tokenized ECE.

I didn't want to leave the `wc_ajax_wcpay_add_to_cart` handler for the tokenized ECE if it wasn't going to be needed.

htdat reviewed (changes requested):
I tested with the checkout and settings pages, they're all working well.

However, I am seeing the `We're not able to process this payment. Please refresh the page and try again` when trying to use Google Pay in a product page. Looking at the logs, there is no request sent to WPCOM so I guess something wrong in the AJAX changes. 

Also, I think the part of moving AJAX actions around would be in a follow-up PR. Currently, it blows up this PR a bit, and it's not a part of the original requirements in #9725

frosso commented:
Thank you for reviewing again @htdat !

> However, I am seeing the We're not able to process this payment. Please refresh the page and try again when trying to use Google Pay in a product page. Looking at the logs, there is no request sent to WPCOM so I guess something wrong in the AJAX changes.

I am not able to reproduce this issue. I wonder if somehow there are some cached assets?
Right after clicking on the Google Pay/Apple Pay button from the product page, do you see a request to `/wp-json/wc/store/v1/cart/add-item` ? If, instead, you see a request to a `admin-ajax` endpoint with an `ece_*` action, it means that there's some caching issue going on.

Since there are changes to the webpack configuration, maybe try also restarting your `npm run start` process?

> Also, I think the part of moving AJAX actions around would be in a follow-up PR. Currently, it blows up this PR a bit, and it's not a part of the original requirements in https://github.com/Automattic/woocommerce-payments/issues/9725

I understand that moving the methods around was not called out in the ticket's requirements, but it was because I didn't realize it was needed when I wrote the ticket.

Here's my thought process - part of the requirements were:
> All legacy code associated with old implementations is removed [...] No longer used endpoints

So I removed the handlers defined in `includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php`.

I then realized that the `normalize_state` method (which was present in `includes/express-checkout/class-wc-payments-express-checkout-button-helper.php` ) had specific logic to handle HK addresses, and it was not part of the Tokenized ECE. So I decided to port it over in this PR. I think it might have been a bug in the Tokenized ECE. So I would have needed to refactor it in `includes/express-checkout/class-wc-payments-express-checkout-button-helper.php` in any case.

I didn't think that those `get_shipping_options` / `maybe_restore_recurring_chosen_shipping_methods` / `get_normalized_state` / `validate_state`/ etc. methods should have stayed in the helper class (they are specific and only used by the AJAX handlers), so I moved them from `includes/express-checkout/class-wc-payments-express-checkout-button-helper.php` to `includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php`. 

So, since I would have needed to refactor the HK-specific logic and also move the helper methods into the AJAX handler class, I did it all in this PR.

I can move back some of those "address normalization" methods into the `includes/express-checkout/class-wc-payments-express-checkout-button-helper.php` file, but I would still need to refactor them to handle HK addresses.

Please let me know your preference!

frosso committed:
Merge branch 'develop' into chore/remove-deprecated-non-tokenized-ece

htdat commented on includes/express-checkout/class-wc-payments-express-checkout-button-handler.php:
```diff
@@ -92,7 +92,6 @@ public function init() {
 		add_filter( 'woocommerce_registration_redirect', [ $this, 'get_login_redirect_url' ], 10, 3 );
 		add_filter( 'woocommerce_cart_needs_shipping_address', [ $this, 'filter_cart_needs_shipping_address' ], 11, 1 );
 		add_action( 'wp_enqueue_scripts', [ $this, 'scripts' ] );
-		add_action( 'before_woocommerce_pay_form', [ $this, 'display_pay_for_order_page_html' ], 1 );
```
Looking at `display_pay_for_order_page_html`, it seems that we provide data for `wcpayECEPayForOrderParams`. However, `wcpayECEPayForOrderParams` is not used on the client. 

I think it's the correct thing to do but want to double check on this one.

frosso replied:
Yep, you assessment is accurate 👍

htdat commented:
@frosso - thanks for your detailed reply! 

> I am not able to reproduce this issue. I wonder if somehow there are some cached assets?

It is an issue with my testing environment on Pressable due to something with Fraud service. However, it's working well in my local env exposed to the public.

htdat commented:
> So I removed the handlers defined in `includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php`.

This part is clear when I am reviewing this PR. 

> I then realized that the `normalize_state` method (which was present in `includes/express-checkout/class-wc-payments-express-checkout-button-helper.php` ) had specific logic to handle HK addresses, and it was not part of the Tokenized ECE. So I decided to port it over in this PR. I think it might have been a bug in the Tokenized ECE. So I would have needed to refactor it in `includes/express-checkout/class-wc-payments-express-checkout-button-helper.php` in any case.

If I understand correctly, this could have been a bug for Tokenized ECE for a while. IMO, this is more about enhancing for Tokenized, and it is not necessary as a part of this clean-up. It can be addressed in a separate PR. 

> I didn't think that those `get_shipping_options` / `maybe_restore_recurring_chosen_shipping_methods` / `get_normalized_state` / `validate_state`/ etc. methods should have stayed in the helper class (they are specific and only used by the AJAX handlers), so I moved them from `includes/express-checkout/class-wc-payments-express-checkout-button-helper.php` to `includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php`. 

This can also be a separate PR, IMHO. 

--- 

I could understand the main goal behind the original issue #9725 was to remove as much of the dead code as possible. However, given the current size of this PR and the collective knowledge of Gamma regarding this area relatively low, breaking down this PR into smaller ones will be more beneficial. That also allows PR reviewers to focus on and pay attention to each PR.

htdat commented on includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php:
```diff
@@ -55,303 +49,17 @@ public function init() {
-			}
-			wp_send_json( [ 'error' => wp_strip_all_tags( $e->getMessage() ) ], 500 );
-		}
+		add_filter( 'rest_pre_dispatch', [ $this, 'tokenized_cart_store_api_address_normalization' ], 10, 3 );
+		add_filter( 'woocommerce_get_country_locale', [ $this, 'modify_country_locale_for_express_checkout' ], 20 );
```
Nitpick: With this clean-up, these are two remaining filters, then the name `AJAX_Handler` seems no longer relevant.

frosso replied:
I think that the `AJAX_Handler` name is still relevant, because these filters will only be called on AJAX calls. The `WC_Payments_Express_Checkout_Ajax_Handler` class also contains the `ajax_add_to_cart` handler.

What naming did you have in mind for this class, instead?

htdat replied:
Only the `ajax_add_to_cart` method in this class is really to handle AJAX calls. 
All other methods are focusing on the Store API request. So something more specific like `*_Rest_Store_Api_Handler` would be better.

htdat replied:
Also, `ajax_add_to_cart` here is now only to be used with WooPay: 

https://github.com/Automattic/woocommerce-payments/blob/88f7abbce85229e3978c60773a12c0f75f20e677/includes/express-checkout/class-wc-payments-express-checkout-button-display-handler.php#L87-L89

I think we can move this `ajax_add_to_cart` to `WC_Payments_WooPay_Button_Handler` and called it in `WC_Payments_WooPay_Button_Handler::init()`, and remove the condition above.

frosso replied:
> Only the `ajax_add_to_cart` method in this class is really to handle AJAX calls. 
All other methods are focusing on the Store API request. So something more specific like `*_Rest_Store_Api_Handler` would be better. 

I see what you're saying. I was instead thinking that Store API requests are a kind of AJAX requests.

I'll address the rename separately, because I am afraid there's the potential of creating issues with the upgrade. I'd like to get this in before the weekend for GS testing.

frosso committed:
PR feedback: reduce scope of changes

frosso commented:
@htdat I removed the HK address handling in 88f7abb to break down the PR as you suggested. Now the scope should be better coupled to the ticket's description. I'll open a new PR with those changes once this is merged.

htdat reviewed (approved):
Thanks for addressing feedback. 

✅ Code changes are now more simple and easier to follow. 
✅ Manual tests look good: 

- Pay for order / checkout / cart / product pages: WooPay + Google Pay 
- Settings look good. Confirmed that the current state of this PR is still fixing this regression issue https://github.com/Automattic/woocommerce-payments/pull/10761#issuecomment-2903244027

❓ There is still a nitpick comment https://github.com/Automattic/woocommerce-payments/pull/10784#discussion_r2115051194 - feel free to address it or iterate it later as it's really nitpick. Even changes addressing it is minimal. 

🚢 Pre-approve the PR!
